man/csvcut.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "CSVCUT" "1" "Aug 16, 2024" "2.0.1" "csvkit"
.SH NAME
csvcut \- csvcut Documentation
.SH DESCRIPTION
.sp
Filters and truncates CSV files. Like the Unix \(dqcut\(dq command, but for tabular data:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: csvcut [\-h] [\-d DELIMITER] [\-t] [\-q QUOTECHAR] [\-u {0,1,2,3}] [\-b]
              [\-p ESCAPECHAR] [\-z FIELD_SIZE_LIMIT] [\-e ENCODING] [\-S] [\-H]
              [\-K SKIP_LINES] [\-v] [\-l] [\-\-zero] [\-V] [\-n] [\-c COLUMNS]
              [\-C NOT_COLUMNS] [\-x]
              [FILE]

Filter and truncate CSV files. Like the Unix \(dqcut\(dq command, but for tabular
data.

positional arguments:
  FILE                  The CSV file to operate on. If omitted, will accept
                        input as piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-n, \-\-names           Display column names and indices from the input CSV
                        and exit.
  \-c COLUMNS, \-\-columns COLUMNS
                        A comma\-separated list of column indices, names or
                        ranges to be extracted, e.g. \(dq1,id,3\-5\(dq. Defaults to
                        all columns.
  \-C NOT_COLUMNS, \-\-not\-columns NOT_COLUMNS
                        A comma\-separated list of column indices, names or
                        ranges to be excluded, e.g. \(dq1,id,3\-5\(dq. Defaults to no
                        columns. Ignores unknown columns.
  \-x, \-\-delete\-empty\-rows
                        After cutting, delete rows which are completely empty.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See also: \fI\%Arguments common to all tools\fP\&.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
csvcut does not implement row filtering, for this you should pipe data to \fI\%csvgrep\fP\&.
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
If a data row is longer than the header row, its additional columns are truncated.
.UNINDENT
.UNINDENT
.SH EXAMPLES
.SS Print columns
.sp
Print the indices and names of all columns:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ csvcut \-n examples/realdata/FY09_EDU_Recipients_by_State.csv
  1: State Name
  2: State Abbreviate
  3: Code
  4: Montgomery GI Bill\-Active Duty
  5: Montgomery GI Bill\- Selective Reserve
  6: Dependents\(aq Educational Assistance
  7: Reserve Educational Assistance Program
  8: Post\-Vietnam Era Veteran\(aqs Educational Assistance Program
  9: TOTAL
 10:
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Print only the names of all columns, by removing the indices with the \fBcut\fP command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ csvcut \-n examples/realdata/FY09_EDU_Recipients_by_State.csv | cut \-c6\-
State Name
State Abbreviate
Code
Montgomery GI Bill\-Active Duty
Montgomery GI Bill\- Selective Reserve
Dependents\(aq Educational Assistance
Reserve Educational Assistance Program
Post\-Vietnam Era Veteran\(aqs Educational Assistance Program
TOTAL
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Extract columns
.sp
Extract the first and third columns:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvcut \-c 1,3 examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Extract columns named \(dqTOTAL\(dq and \(dqState Name\(dq (in that order):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvcut \-c TOTAL,\(dqState Name\(dq examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Extract a column that may not exist in all files:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo d, | csvjoin examples/dummy.csv \- | csvcut \-c d
 echo d, | csvjoin examples/join_no_header_row.csv \- | csvcut \-c d
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Other
.sp
Add line numbers to a file, making no other changes:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvcut \-l examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Display a column\(aqs unique values:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvcut \-c 1 examples/realdata/FY09_EDU_Recipients_by_State.csv | sed 1d | sort | uniq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Or:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvcut \-c 1 examples/realdata/FY09_EDU_Recipients_by_State.csv | csvsql \-\-query \(aqSELECT DISTINCT(\(dqState Name\(dq) FROM stdin\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/csvcut.py
#!/usr/bin/env python

"""
csvcut is originally the work of eminent hackers Joe Germuska and Aaron Bycoffe.

This code is forked from:
https://gist.github.com/561347/9846ebf8d0a69b06681da9255ffe3d3f59ec2c97

Used and modified with permission.
"""

import sys

import agate

from csvkit.cli import CSVKitUtility


class CSVCut(CSVKitUtility):
    description = 'Filter and truncate CSV files. Like the Unix "cut" command, but for tabular data.'
    override_flags = ['L', 'I']

    def add_arguments(self):
        self.argparser.add_argument(
            '-n', '--names', dest='names_only', action='store_true',
            help='Display column names and indices from the input CSV and exit.')
        self.argparser.add_argument(
            '-c', '--columns', dest='columns',
            help='A comma-separated list of column indices, names or ranges to be extracted, e.g. "1,id,3-5". '
                 'Defaults to all columns.')
        self.argparser.add_argument(
            '-C', '--not-columns', dest='not_columns',
            help='A comma-separated list of column indices, names or ranges to be excluded, e.g. "1,id,3-5". '
                 'Defaults to no columns.')
        self.argparser.add_argument(
            '-x', '--delete-empty-rows', dest='delete_empty', action='store_true',
            help='After cutting, delete rows which are completely empty.')

    def main(self):
        if self.args.names_only:
            self.print_column_names()
            return

        if self.additional_input_expected():
            sys.stderr.write('No input file or piped data provided. Waiting for standard input:\n')

        rows, column_names, column_ids = self.get_rows_and_column_names_and_column_ids(**self.reader_kwargs)

        output = agate.csv.writer(self.output_file, **self.writer_kwargs)
        output.writerow([column_names[column_id] for column_id in column_ids])

        for row in rows:
            out_row = [row[column_id] if column_id < len(row) else None for column_id in column_ids]

            if not self.args.delete_empty or any(out_row):
                output.writerow(out_row)


def launch_new_instance():
    utility = CSVCut()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
